### ISSUE TITLE
unconditional_panic lint getting recognized in code that actually panics condtionally

### ISSUE THREAD

**Issue by pro465** [rust-lang/rust#90534]
I tried this code:

```rust
fn main() {
    let v: [i32; 0] = [];
    let _ = v.len() > 0 && v[0] == 0;
}
```

I expected to see this happen: *code compiles and runs fine, without any error or panic*

Instead, this happened: 
```
error: this operation will panic at runtime
 --> t.rs:3:28
  |
3 |     let _ = v.len() > 0 && v[0] == 0;
  |                            ^^^^ index out of bounds: the length is 0 but the index is 0
  |
  = note: `#[deny(unconditional_panic)]` on by default
```

### Meta

`rustc --version --verbose`:
```
rustc 1.56.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: armv7-linux-androideabi
release: 1.56.0
LLVM version: 13.0.0
```

**Comment by JanBeh** [comment]
Also note that this triggers the lint:

```
fn main() {
    let xs: [i32; 5] = [1, 2, 3, 4, 5];
    if false {
        println!("{}", xs[7]);
    }
}
```

(using `rustc 1.63.0-nightly (43347397f 2022-06-23)`)

Example [taken from URLO](https://users.rust-lang.org/t/out-of-bound-compile-error-with-len-function-call/77342/8?u=jbe).

**Comment by JanBeh** [comment]
See also: https://github.com/rust-lang/rust/issues/78803

**Comment by pro465** [comment]
@JanBeh thanks for the update! hopefully somebody will notice that issue and prepare a fix

**Comment by Nilstrieb** [comment]
Closing in favor of #78803

**Comment by pro465** [comment]
@Nilstrieb , considering that this issue is still there in the latest nightly, even after #78803 was fixed, should we reopen this?

**Comment by ahqsoftwares** [comment]
There's something to note as well, it should be converted to a warning cuz the conditional part would never get called
